File: tests/python/src/testcancel.py
import sys
import time
import threading
import cameopy

def cancelAll():
    time.sleep(1)
    cameopy.This.cancelAll()

def cancelWaitFor(instance):
    time.sleep(1)
    instance.cancel()

def cancelWaitForSubscribers(publisher):
    time.sleep(1)
    publisher.cancel()

def killApplication(app):
    time.sleep(1)
    app.kill()

def cancelResponder(responder):
    time.sleep(1)
    responder.cancel()

def loopResponder(responder):
    
    while not responder.isCanceled():
        responder.receive()
    

def cancelRequester(requester, responder):
    time.sleep(1)
    requester.cancel()
    responder.cancel()


this = cameopy.This
this.init(sys.argv)

useProxy = False if len(sys.argv) < 3 else (sys.argv[1] == "true") 

options = 0
endpoint = "tcp://localhost:11000";
if useProxy:
    options |= cameopy.option.USE_PROXY
    endpoint = "tcp://localhost:12000";

server = cameopy.Server.create(endpoint, options)
server.init()

def testCancelAll():

    print("Starting stopcpp for cancelAll")
    
    app = server.start("stopcpp")
    
    # Start thread.
    t = threading.Thread(target=cancelAll)
    t.start()
    
    app.waitFor()
    
    print("End of waitFor")
    
    app.stop()
    state = app.waitFor()
    
    t.join()
    
    print("End of stopcpp with state", cameopy.toString(state))
    
    app.terminate()


def testCancelWaitFor():

    print("Starting stopcpp for cancelWaitFor")
    
    app = server.start("stopcpp")
    
    # Start thread.
    t = threading.Thread(target=cancelWaitFor, args=(app,))
    t.start()
    
    app.waitFor()
    
    print("End of waitFor")
    
    app.stop()
    state = app.waitFor()
    
    t.join()
    
    print("End of stopcpp with state", cameopy.toString(state))
    
    app.terminate()

    
def testCancelWaitForSubscribers():

    print("Creating publisher and waiting for 1 subscriber...")

    publisher = cameopy.coms.Publisher.create("publisher")
    publisher.setWaitForSubscribers(1)

    # Start thread.
    t = threading.Thread(target=cancelWaitForSubscribers, args=(publisher,))
    t.start()

    publisher.init()

    t.join()

    print("Synchronization with the subscriber", not publisher.isCanceled())
    
    publisher.terminate()


def testKillApplication():
        
    print("Starting publisherlooppy for killing")

    app = server.start("publisherlooppy")

    # Start kill thread.
    t = threading.Thread(target=killApplication, args=(app,))
    t.start()

    subscriber = cameopy.coms.Subscriber.create(app, "publisher")
    subscriber.setCheckApp(True)
    subscriber.init()

    # Receiving data.
    while True:
        data = subscriber.receiveString()
        if data:
            print("Received", data)
        else:
            break

    print("Subscriber end of stream", subscriber.hasEnded())

    state = app.waitFor()

    print("End of publisherlooppy with state", cameopy.toString(state))

    t.join()
    
    subscriber.terminate()
    app.terminate()
    

def testCancelSubscriber():
    
    print("Starting publisherlooppy for testing cancel of a subscriber")

    app = server.start("publisherlooppy")

    # Create a subscriber.
    subscriber = cameopy.coms.Subscriber.create(app, "publisher")
    subscriber.setCheckApp(True)
    subscriber.init()
    
    # Start cancel thread.
    t = threading.Thread(target=cancelAll)
    t.start()

     # Receiving data.
    while True:
        data = subscriber.receiveString()
        if data:
            print("Received", data)
        else:
            break

    print("Subscriber end of stream", subscriber.hasEnded())
    
     # Start kill thread.
    k = threading.Thread(target=killApplication, args=(app,))
    k.start()

    state = app.waitFor()

    print("End of publisherloopcpp with state", cameopy.toString(state))

    t.join()
    k.join()
    
    subscriber.terminate()
    app.terminate()


def testResponder():
    
    print("Creating basic responder and waiting for requests")
    
    responder = cameopy.coms.basic.Responder.create("responder")
    responder.init()
    
    # Start cancel thread.
    t = threading.Thread(target=cancelResponder, args=(responder,))
    t.start()

    print("Wait for requests")

    request = responder.receive()
    
    if request is not None:
        print("Responder error: receive should return None")

    t.join()
    
    responder.terminate()
    

def testRequester():
    
    print("Creating basic responder and requester")
    
    responder = cameopy.coms.basic.Responder.create("responder")
    responder.init()
    
    # Start reponder thread.
    tr = threading.Thread(target=loopResponder, args=(responder,))
    tr.start()
    
    thisApp = server.connect(cameopy.This.getName())
    
    requester = cameopy.coms.Requester.create(thisApp, "responder")
    requester.init()
    
    # Start cancel thread.
    tc = threading.Thread(target=cancelRequester, args=(requester, responder,))
    tc.start()

    print("Sending request")

    requester.send("request")

    print("Receiving response")
    
    requester.receive()
    
    if requester.isCanceled():
        print("Requester is canceled")
    else:
        print("Requester is not canceled")

    tc.join()
    tr.join()

    requester.terminate()    
    responder.terminate()

    
testCancelAll()
testCancelWaitFor()
testCancelWaitForSubscribers()
testKillApplication()
testCancelSubscriber()
testResponder()
testRequester()

server.terminate()
this.terminate()

print("Finished the application")

File: tests/cpp/src/TestCancel.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include <cameo/api/cameo.h>
#include <thread>
#include <iostream>

using namespace std;
using namespace cameo;

int main(int argc, char *argv[]) {

	This::init(argc, argv);

	int options = 0;
	bool useProxy = false;
	string endpoint = "tcp://localhost:11000";
	if (argc > 2) {
		useProxy = (string(argv[1]) == "true");
	}
	if (useProxy) {
		options |= USE_PROXY;
		endpoint = "tcp://localhost:10000";
	}

	unique_ptr<Server> server = Server::create(endpoint, options);
	server->init();

	// Test the cancelAll function.
	{
		cout << "Starting stopcpp for cancelAll" << endl;

		unique_ptr<App> stopApplication = server->start("stopcpp");

		// Start thread.
		thread cancelThread([] {
			this_thread::sleep_for(chrono::seconds(1));
			This::cancelAll();
		});

		State state = stopApplication->waitFor();

		cout << "End of waitFor with state " << toString(state) << endl;

		stopApplication->stop();
		state = stopApplication->waitFor();

		cout << "End of stopcpp with state " << toString(state) << endl;

		cancelThread.join();
	}

	// Test the cancel function.
	{
		cout << "Starting stopcpp for cancelWaitFor" << endl;

		// Use a shared_ptr to use it in the thread and the main thread.
		shared_ptr<App> stopApplication(server->start("stopcpp"));

		// Start thread.
		thread cancelThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			stopApplication->cancel();
		});

		State state = stopApplication->waitFor();

		cout << "End of waitFor with state " << toString(state) << endl;

		stopApplication->stop();
		state = stopApplication->waitFor();

		cout << "End of stopcpp with state " << toString(state) << endl;

		cancelThread.join();
	}

	// Test cancel
	{
		cout << "Creating publisher and waiting for 1 subscriber..." << endl;

		// Use a shared_ptr to use it in the thread and the main thread.
		shared_ptr<coms::Publisher> publisher(coms::Publisher::create("publisher"));
		publisher->setWaitForSubscribers(1);

		// Start thread.
		thread cancelThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			publisher->cancel();
		});

		publisher->init();

		cancelThread.join();

		cout << "Synchronization with subscriber " << !publisher->isCanceled() << endl;
	}

	// Test the killing of the application.
	{
		cout << "Starting publisherloopcpp for killing" << endl;

		// Use a shared_ptr to use it in the thread and the main thread.
		shared_ptr<App> pubLoopApplication(server->start("publisherloopcpp"));

		// Start thread.
		thread killThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			pubLoopApplication->kill();
		});

		// Create a subscriber checking the app.
		unique_ptr<coms::Subscriber> subscriber = coms::Subscriber::create(*pubLoopApplication, "publisher");
		subscriber->setCheckApp(true);
		subscriber->init();

		// Receiving data.
		while (true) {
			optional<string> data = subscriber->receive();
			if (!data.has_value()) {
				break;
			}
			cout << "Received " << data.value() << endl;
		}

		cout << "Subscriber end of stream " << subscriber->hasEnded() << endl;

		State state = pubLoopApplication->waitFor();

		cout << "End of publisherloopcpp with state " << toString(state) << endl;

		killThread.join();
	}

	// Test the Subscriber init.
	{
		cout << "Creating subscriber for being canceled" << endl;

		// Get this app.
		unique_ptr<App> thisApp = server->connect(This::getName());

		// Create a requester.
		unique_ptr<coms::Subscriber> subscriber = coms::Subscriber::create(*thisApp, "an unknown publisher");

		// Start thread.
		thread cancelThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			subscriber->cancel();
		});

		subscriber->init();

		cout << "Subscriber ready and canceled ? " << subscriber->isReady() << " " << subscriber->isCanceled() << endl;

		cancelThread.join();
	}

	// Test the canceling of a subscriber.
	{
		cout << "Starting publisherloopcpp for testing cancel of a subscriber" << endl;

		// Use a shared_ptr to use it in the thread and the main thread.
		shared_ptr<App> pubLoopApplication(server->start("publisherloopcpp"));

		// Create a subscriber.
		unique_ptr<coms::Subscriber> subscriber = coms::Subscriber::create(*pubLoopApplication, "publisher");
		subscriber->init();

		// Start thread.
		thread cancelThread([] {
			this_thread::sleep_for(chrono::seconds(1));
			This::cancelAll();
		});

		// Receiving data.
		while (true) {
			optional<string> data = subscriber->receive();
			if (!data.has_value()) {
				break;
			}
			cout << "Received " << data.value() << endl;
		}

		cout << "Subscriber end of stream " << subscriber->hasEnded() << endl;

		// Start thread.
		thread killThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			pubLoopApplication->kill();
		});

		State state = pubLoopApplication->waitFor();

		cout << "End of publisherloopcpp with state " << toString(state) << endl;

		cancelThread.join();
		killThread.join();
	}

	// Test the basic Responder.
	{
		cout << "Creating basic responder and waiting for requests" << endl;

		// Create a responder.
		unique_ptr<coms::basic::Responder> responder = coms::basic::Responder::create("responder");
		responder->init();

		// Start thread.
		thread cancelThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			responder->cancel();
		});

		cout << "Wait for requests" << endl;

		unique_ptr<coms::basic::Request> request = responder->receive();

		if (request) {
			cerr << "Responder error: receive should return null" << endl;
		}

		cancelThread.join();
	}

	// Test the Requester init.
	{
		cout << "Creating requester for being canceled" << endl;

		// Get this app.
		unique_ptr<App> thisApp = server->connect(This::getName());

		// Create a requester.
		unique_ptr<coms::Requester> requester = coms::Requester::create(*thisApp, "an unknown responder");

		// Start thread.
		thread cancelThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			requester->cancel();
		});

		requester->init();

		cout << "Requester ready and canceled ? " << requester->isReady() << " " << requester->isCanceled() << endl;

		cancelThread.join();
	}

	// Test the basic Requester.
	{
		cout << "Creating basic responder and requester" << endl;

		// Create a responder.
		unique_ptr<coms::basic::Responder> responder = coms::basic::Responder::create("responder");
		responder->init();

		// Start thread.
		thread responderThread([&] {
			while (!responder->isCanceled()) {
				responder->receive();
			}
		});

		// Get this app.
		unique_ptr<App> thisApp = server->connect(This::getName());

		// Create a requester.
		unique_ptr<coms::Requester> requester = coms::Requester::create(*thisApp, "responder");
		requester->init();

		// Start thread.
		thread cancelThread([&] {
			this_thread::sleep_for(chrono::seconds(1));
			requester->cancel();
			responder->cancel();
		});

		cout << "Sending request" << endl;

		requester->send("request");

		cout << "Receiving response" << endl;

		requester->receive();

		if (requester->isCanceled()) {
			cout << "Requester is canceled" << endl;
		}
		else {
			cout << "Requester is not canceled" << endl;
		}

		cancelThread.join();
		responderThread.join();
	}

	// Test the multi responder.
	{
		cout << "Creating multi responder" << endl;

		unique_ptr<coms::multi::ResponderRouter> router = coms::multi::ResponderRouter::create("responder");
		router->init();

		unique_ptr<coms::multi::Responder> responder = coms::multi::Responder::create(*router);
		responder->init();

		std::thread routerThread([&] {
			router->run();
		});

		std::thread responderThread([&] {
			responder->receive();
		});

		responder->cancel();
		router->cancel();

		if (responder->isCanceled() && router->isCanceled()) {
			cout << "Router and responder are canceled" << endl;
		}
		else {
			cout << "Router and responder are not canceled" << endl;
		}

		responderThread.join();
		routerThread.join();
	}

	cout << "Finished the application" << endl;

	return 0;
}
